### ISSUE TITLE
Using my own dataset.

### ISSUE THREAD

**Issue by supercrazysam** [CIA-Oceanix/GeoTrackNet#13]
Dear author, thanks for your awesome work. 

However, I encountered some issue when I tried to use my own dataset.
I would like to make sure that I am doing it correctly, so I will try to verify my steps here with you first.

==========
Here are the steps and changes that I applied, by following some of your replies on other issues.

For csv2pkl.py:
   -  Modified the `LAT_MIN, LAT_MAX, LON_MIN, LON_MAX` according to my dataset,
      ```LAT_MIN = 40.33   LAT_MAX = 40.77   LON_MIN = -74.30   LON_MAX = -73.50 ```
   -  Modified the file names/ path names/ time ranges
   -  Changed the dataset format so that when csv2pkl read my dataset, it will fit in in the format and order of
    `LAT, LON, SOG, COG, HEADING, ROT, NAV_STT, TIMESTAMP, MMSI, SHIPTYPE, D2C`
   -  Zero-ed out ROT, NAV_STT and D2C,  also removed the D2C filter in the code below.
   -  Set CARGO_TANKER_ONLY = False
   -  Removed the UNIX_TIME /1000 stuff

For dataset_preprocessing.py:
   - Modified the LAT_MIN, LAT_MAX, LON_MIN, LON_MAX, dataset+path name

After the modification steps on csv2pkl.py and dataset_preprocessing mentioned above,
using my datasets, I was able to create 3 pkl files (train, test, valid), 
alongside 3 png (train, test, valid) in which I can see the filtered tracks inside the ROI. 

==========

However, after this, I am not sure what is the next step, please help!

1. What is the next step? 
2. For the next step, What are the files that I will need to change? What value should I change it to?
3. Especially since I am using the default trained model
    `chkpt/elbo-ct_2017010203_10_20_train.pkl-data_dim-602-latent_size-100-batch_size-50.zip`,
    do I need to make sure my dataset have the same (LAT_RANGE and LON_RANGE), for the pretrained model to work?

Also some extra question here:
   Does the default pretrained model work well on new area that it was not trained on?
   

Thanks for your help in advance!

**Comment by duonginspace** [comment]
Hello, 

It seems like your data are ready. Now you should follow the step by step procedure presented in the README to train the network. 

`Does the default pretrained model work well on new area that it was not trained on?`
No, AIS data are geographically dependent, a model trained on datasets in 1 ROI won't work in other ROIs.

Good luck!

**Comment by supercrazysam** [comment]
@dnguyengithub Thanks for your lightning-fast response!

I want to make sure that, to continue the process, I just need to change the filenames, LAT, LON params in the commands,
like below? 

Basically, I want to know if there is any other internal code modification required for this to work, before starting the training process.
(Such as any parameter changes in code... etc)

Since my ROI is 
`LAT_MIN = 40.33 LAT_MAX = 40.77 LON_MIN = -74.30 LON_MAX = -73.50 `

============
For training
`python geotracknet.py \
     --mode=train \
     --dataset_dir=./data \
     --trainingset_name=new_dataset_train.pkl \
     --testset_name=new_dataset_valid.pkl \
     --lat_min= 40.33  \
     --lat_max= 40.77 \
     --lon_min=-74.30 \
     --lon_max=-73.50  \
     --latent_size=100 \
     --batch_size=32 \
     --num_samples=16 \
     --learning_rate=0.0003 \`
     
For local_logprob
`python geotracknet.py \
     --mode=local_logprob \
     --dataset_dir=./data \
     --trainingset_name=new_dataset_train.pkl \
     --testset_name=new_dataset_valid.pkl \
     --lat_min= 40.33  \
     --lat_max= 40.77 \
     --lon_min=-74.30 \
     --lon_max=-73.50  \
     --latent_size=100 \
     --batch_size=32 \
     --num_samples=16 \
     --learning_rate=0.0003 \`
     
 For contrario_detection:
 `python geotracknet.py \
      --mode=contrario_detection \
      --dataset_dir=./data \
     --trainingset_name=new_dataset_train.pkl \
     --testset_name=new_dataset_valid.pkl \
     --lat_min= 40.33  \
     --lat_max= 40.77 \
     --lon_min=-74.30 \
     --lon_max=-73.50  \
      --contrario_eps=1e-10 \
      --latent_size=100 \
      --batch_size=32 \
      --num_samples=16 \
      --learning_rate=0.0003 \
`

**Comment by duonginspace** [comment]
Yes, it looks correct to me.

**Comment by supercrazysam** [comment]
Hi @dnguyengithub. After csv2pkl, dataset_preprocessing and calculate_AIS_mean. 
I am currently trying to start training the network, however I encountered some issues.

Issue & Error log:
```
Traceback (most recent call last):
  File "geotracknet.py", line 76, in <module>
    runners.run_train(config)
  File "GeoTrackNet/runners.py", line 316, in run_train
    bound, train_op, global_step = create_graph()
  File "/home/cshum/Desktop/GeoTrackNet/runners.py", line 306, in create_graph
    bound, loss = create_loss()
  File "GeoTrackNet/runners.py", line 278, in create_loss
    repeat=True) 
...
ValueError: Dimensions must be equal, but are 227 and 226 for 'sub' (op: 'Sub') with input shapes: [?,?,227], [1,1,226].
```

==========

Here is what I did
Since I am using my own ROI 
`LAT_MIN = 40.33 LAT_MAX = 40.77 LON_MIN = -74.30 LON_MAX = -73.50 `

In **flags_config.py**, 
I modified `tf.app.flags.DEFINE_float  lat_min,lat_max,lon_min,lon_max`  to my own ROI

In **dataset_preprocessing.py**, 
I modified `parser  --lat_min, --lat_max, --lon_min, --lon_max`  to my own ROI

In **datasets.py**, 
I modified the create_dense_vect, to become 
`create_dense_vect(msg,lat_bins = 44, lon_bins = 80, sog_bins = 30 ,cog_bins = 72):`

In **calculate_AIS_mean.py**, 
I changed the `LAT_BIN, LON_BINS, SOG_BINS, COG_BINS ` value to
`LAT_BINS = 44; LON_BINS = 80; SOG_BINS = 30; COG_BINS = 72`
I also modified the create_dense_vect, to become 
`create_dense_vect(msg,lat_bins = 44, lon_bins = 80, sog_bins = 30 ,cog_bins = 72):`

==========
The training command that I ran, that caused the error above:

`python geotracknet.py \ --mode=train \ --dataset_dir=./data \ --trainingset_name=dataset_train.pkl \ --testset_name=dataset_test.pkl \ --lat_min= 40.33 \ --lat_max= 40.77 \ --lon_min=-74.30 \ --lon_max=-73.50 \ --latent_size=100 \ --batch_size=32 \ --num_samples=16 \ --learning_rate=0.0003 \`

I also changed the latent_size to 100 in `flags_config.py`, in order to match this command.

==========
Did I miss any detail? Did I made mistake? 
Your help will be very useful to me, thanks!

**Comment by duonginspace** [comment]
Hello, 

This is an error related to the dimensions of your 4-hot vectors. With your setting, the dimension of the 4-hot vectors should be 226.
Could you check which one has the dimension of 227?

**Comment by supercrazysam** [comment]
@dnguyengithub thanks for the reminder, I fixed the mentioned problem, it turn out that the math.ceil for bins in flags_config.py made 226 become 227.

==========
However, I encountered another problem when I am trying to run detection using my trained model.

While I am doing contrario_detection, the result prints

```
Global step:  80001
100%|███████████████████████| 4399/4399 [36:04<00:00,  2.03it/s]
Number of processed tracks:  4399
Number of abnormal tracks:  15
Number of errors:  2453
```

Is it normal to have such large number of errors? I don't remember seeing any error when running the demo.

The command used was:
```
python3 geotracknet.py \
  --mode=contrario_detection \
  --dataset_dir=./AIS_processed \
  --trainingset_name=AIS_20180101_valid_track.pkl \
  --testset_name=AIS_20180101_test_track.pkl \
  --lat_min=40.33 \
  --lat_max=40.77 \
  --lon_min=-74.30 \
  --lon_max=-73.50 \
  --contrario_eps=1e-10 \
  --latent_size=100 \
  --batch_size=32 \
  --num_samples=16 \
  --learning_rate=0.0003 \
```
==========
Something that I have tried:

1. I bypassed the try except in contrario_detection to see what went wrong, and discovered that the error produced was:
```
File "geotracknet.py", line 343, in <module>
    l_local_log_prod = Map_logprob[str(d_row)+","+str(d_col)]
KeyError: '7,8'
```

==========
After that, I tried to print out some variables to check their values when the KeyError: '7,8' occurs

For **Map_logprob**
The key list for Map_logprob is
```
dict_keys(['0,0', '0,1', '0,2', '0,3', '0,4', '0,5', '0,6', '0,7', '1,0', '1,1', '1,2', '1,3', '1,4', '1,5', '1,6', '1,7', '2,0', '2,1', '2,2', '2,3', '2,4', '2,5', '2,6', '2,7', '3,0', '3,1', '3,2', '3,3', '3,4', '3,5', '3,6', '3,7', '4,0', '4,1', '4,2', '4,3', '4,4', '4,5', '4,6', '4,7'])
```
 
 For **tmp**
 The shape of tmp is (100, 4)
 The value of tmp is:
```
[[ 37  76 124 139]
 [ 37  76 124 182]
 [ 37  76 124 131]
 [ 40  77 146 155]
 [ 43  79 141 154]
 [ 43  79 122 141]
 [ 42  78 132 157]
 [ 41  77 118 130]
 [ 38  76 145 193]
 [ 37  73 124 152]
 [ 38  77 140 158]
 [ 41  76 124 166]
 [ 42  78 133 158]
 [ 43  79 133 161]
 [ 43  79 125 140]
 [ 42  78 129 161]
 [ 41  76 125  37]
 [ 75 124 152  37]
 [ 73 124 136  38]
 [ 77 144 157  41]
 [ 76 124 175  42]
 [ 78 134 156  43]
 [ 79 142 161  43]
 [ 79 122 143  42]
 [ 78 125 177  41]
 [ 76 127 138  38]
 [ 76 148 194  37]
 [ 73 124 165  38]
 [ 76 146 158  41]
 [ 76 124 167  42]
 [ 78 136 159  43]
 [ 79 133 161  43]
 [ 79 126 140  42]
 [ 78 127 175  41]
 [ 76 124 139  38]
 [ 76 144 146  37]
 [ 73 124 167  39]
 [ 76 145 155  41]
 [ 76 124 138  42]
 [ 78 133 157  43]
 [ 79 137 161  43]
 [ 79 122 142  42]
 [ 78 127 178  41]
 [ 76 127 137  38]
 [ 76 123 146  37]
 [ 73 124 172  38]
 [ 76 145 157  41]
 [ 76 124 152  41]
 [ 78 124 167  43]
 [ 79 138 160  43]
 [ 79 126 140  42]
 [ 78 128 177  41]
 [ 77 116 129  38]
 [ 76 117 146  37]
 [ 73 124 140  39]
 [ 77 149 156  41]
 [ 76 124 167  41]
 [ 78 132 160  43]
 [ 79 134 159  43]
 [ 79 129 136  42]
 [ 78 124 193  41]
 [ 77 121 132  39]
 [ 77 145 191  37]
 [ 73 124 159  38]
 [ 76 147 159  41]
 [ 76 124 172  41]
 [ 78 131 149  43]
 [ 79 134 160  43]
 [ 79 128 137  42]
 [ 78 127 169  41]
 [ 77 125 134  39]
 [ 77 149 192  37]
 [ 73 124 174  38]
 [ 76 143 163  41]
 [ 76 124 193  41]
 [ 78 131 157  43]
 [ 79 133 160  43]
 [ 80 129 138  42]
 [ 78 126 186  41]
 [ 77 129 133  39]
 [ 76 137 192  37]
 [ 73 124 144  38]
 [ 76 145 155  41]
 [ 76 124 126  41]
 [ 78 129 150  42]
 [ 78 132 161  43]
 [ 79 129 137  42]
 [ 78 127 192  41]
 [ 77 127 133  38]
 [ 76 148 192  37]
 [ 73 124 178  38]
 [ 76 143 163  41]
 [ 76 124 135  41]
 [ 78 128 158  43]
 [ 79 133 161  43]
 [ 79 130 142  42]
 [ 78 124 159  41]
 [ 78 125 137  41]
 [ 76 124 172  37]
 [ 75 124 151  37]]
```
==========
So it seems that some value in the tmp list, would calculate a `d_row, d_col` that does not exist in the `Map_logprob` key list.
I have no idea what caused this problem, your help will be greatly appreciated

**Comment by duonginspace** [comment]
Hi, 
Its not normal that you have such large number of errors. 
Did you set the correct value for `cell_lat_reso` and `cell_lon_reso`?

Assuming that you used the setting in the previous question, your `Map_lobrob` should be 4x7. Why do you have key (7,8)?

**Comment by supercrazysam** [comment]
@dnguyengithub,  
- I did use the setting in the previous question.
- I didn't change the default value of `cell_lat_reso` and `cell_lon_reso` value in flags_config.py, they are both 0.1. What value of these two variables should I choose? 
-  There is no key (7,8) in `Map_logprob`, it is just that I found out the program is looking for key (7,8), based on the value in `tmp`
- For `Map_logprob`, in the current situation, it gives out a 5 for `config.n_lat_cells`. For example the `config.n_lat_cells` will be ((40.77-40.33) / 0.1) = 4.4, after math.ceil, it gives 5.  
- So basically, for `LAT_BINS = 44; LON_BINS = 80`, the `Map_logprob` should be 4x7? If this is the case, after this I will use this value as reference to go through the code.

**Comment by duonginspace** [comment]
`Map_logprob` should be 5x8, sorry, my bad.

Could you check why you have (7,8) in your code? Note that I hard-code `config.onehot_lat/lon_reso=0.01` in lines 228, 229.

Sorry but I'm currently not using this version of GeoTrackNet, I cant re-produce the bug to help you.

**Comment by supercrazysam** [comment]
@dnguyengithub 

The `config.onehot_lat/lon_reso=0.01` lines 228, 229, are consistent with my code, since the default value for the one_hot resolution for both lat, lon is also default's 0.01 .

I was able to track the source of key (7,8) error in the code, it actually appears as early as running the `local_logprob` part.
When I uncomment the try except part, the row, col part gives the exact same error:
`KeyError: '7,8'`.  

In both `local_logprob` and `contrario_detection` part, to calculate col and row, eventually it will take in a element of `tmp`, 
with the value of `[ 75 124 152  37]`, 
it will gives int(75*0.01/0.1) = 7 for row,  
gives int((124 - 44)*0.01/0.1) =8 for col.
Thus causing `KeyError: '7,8'` 

Since the row, col was calculated using `tmp[d_timestep,0]` and `tmp[d_timestep,1]`.  
This make me believe that the value in D["seq"] (basically tmp) is what caused the problem.

However, D["seq"]'s array value is already coming from tar, the sess.run() inference output. I am not sure what went wrong.


==========
Another thing that I should mention, and might be related to solving this issue:
During the `save_logprob` stage, there was another fatal error:
```
Traceback (most recent call last):
  File "geotracknet.py", line 188, in <module>
    D["seq"] = np.nonzero(tar[:seq_len_d,d_idx_inbatch,:])[1].reshape(-1,4)
ValueError: cannot reshape array of size 403 into shape (4)
```
And I found out that the length of `np.nonzero(tar[:seq_len_d,d_idx_inbatch,:])[1]` is not always divisible by 4, thus causing error when reshape(-1,4).  (But I don't see this problem when running the demo)

I attempted to fix this by trimmed off the remainder length, making sure it is always divisible by 4, using
```
temp_length = len(np.nonzero(tar[:seq_len_d,d_idx_inbatch,:])[1])
D["seq"] = np.nonzero(tar[:seq_len_d,d_idx_inbatch,:])[1][:((temp_length//4)*4)].reshape(-1,4)
```

Could this be related to the key error? but trimming off the extra remainder length of the array, should not affect the value of the array.

==========
I forked GeoTrackNet and updated it with my modifications, might be helpful:
https://github.com/supercrazysam/GeoTrackNet

 
 Again, thank you so much for your time and help.

**Comment by lovelyyunamars** [comment]
> @dnguyengithub
> 
> The `config.onehot_lat/lon_reso=0.01` lines 228, 229, are consistent with my code, since the default value for the one_hot resolution for both lat, lon is also default's 0.01 .
> 
> I was able to track the source of key (7,8) error in the code, it actually appears as early as running the `local_logprob` part.
> When I uncomment the try except part, the row, col part gives the exact same error:
> `KeyError: '7,8'`.
> 
> In both `local_logprob` and `contrario_detection` part, to calculate col and row, eventually it will take in a element of `tmp`,
> with the value of `[ 75 124 152 37]`,
> it will gives int(75*0.01/0.1) = 7 for row,
> gives int((124 - 44)*0.01/0.1) =8 for col.
> Thus causing `KeyError: '7,8'`
> 
> Since the row, col was calculated using `tmp[d_timestep,0]` and `tmp[d_timestep,1]`.
> This make me believe that the value in D["seq"] (basically tmp) is what caused the problem.
> 
> However, D["seq"]'s array value is already coming from tar, the sess.run() inference output. I am not sure what went wrong.
> 
> ==========
> Another thing that I should mention, and might be related to solving this issue:
> During the `save_logprob` stage, there was another fatal error:
> 
> ```
> Traceback (most recent call last):
>   File "geotracknet.py", line 188, in <module>
>     D["seq"] = np.nonzero(tar[:seq_len_d,d_idx_inbatch,:])[1].reshape(-1,4)
> ValueError: cannot reshape array of size 403 into shape (4)
> ```
> 
> And I found out that the length of `np.nonzero(tar[:seq_len_d,d_idx_inbatch,:])[1]` is not always divisible by 4, thus causing error when reshape(-1,4). (But I don't see this problem when running the demo)
> 
> I attempted to fix this by trimmed off the remainder length, making sure it is always divisible by 4, using
> 
> ```
> temp_length = len(np.nonzero(tar[:seq_len_d,d_idx_inbatch,:])[1])
> D["seq"] = np.nonzero(tar[:seq_len_d,d_idx_inbatch,:])[1][:((temp_length//4)*4)].reshape(-1,4)
> ```
> 
> Could this be related to the key error? but trimming off the extra remainder length of the array, should not affect the value of the array.
> 
> ==========
> I forked GeoTrackNet and updated it with my modifications, might be helpful:
> https://github.com/supercrazysam/GeoTrackNet
> 
> Again, thank you so much for your time and help.

Hi, Sam,

I encountered same problems

Traceback (most recent call last):
  File "geotracknet.py", line 188, in <module>
    D["seq"] = np.nonzero(tar[:seq_len_d,d_idx_inbatch,:])[1].reshape(-1,4)
ValueError: cannot reshape array of size 403 into shape (4)

Also do not now why this happened

Did you solve this problem?

**Comment by supercrazysam** [comment]
Hi @lovelyyunamars , glad to see someone was able to follow the same footsteps and reach this stage.
However, after my analysis on the problem (that you quoted), I was stuck and still not able to solve this issue.
My focus changed to other project after that.

Wish you good luck on solving this problem, please let me know when you found a solution! Thanks.

**Comment by lovelyyunamars** [comment]
Hey, Sam,


I changed the code you provided, and it works for my datasets,&nbsp;then it works for my datasets, I just used on month data, 20 days for training, 8 days for validation, 3 days for testing. there are 6 tracks after prepcoessing, 1 error, 1 abnormal, 4 normal.


Also I found another way to make the code work for my datasets, in datasets_preprocessing.py, I changed the resample resolution 5 mins to 10 mins only for test dataset only, the tracks changed to 3 from 6, then I do not need to do any change in geotracknet.py.


So in my opinion, if we use different datasets with the demo, several parameters in datasets_preprocessing.py need to be modified according to our own datasets.


If you come back to this project agian, and fingure out deep reasons, let me know;）


Regards
Yu


temp_length = len(np.nonzero(tar[:seq_len_d,d_idx_inbatch,:])[1]) D["seq"] = np.nonzero(tar[:seq_len_d,d_idx_inbatch,:])[1][:((temp_length//4)*4)].reshape(-1,4)



 




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "CIA-Oceanix/GeoTrackNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年4月1日(星期四) 中午1:52
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [CIA-Oceanix/GeoTrackNet] Using my own dataset. (#13)





 
Hi lovelyyunamars, glad to see someone was able to follow the same footsteps and reach this stage.
 However, after my analysis on the problem (that you quoted), I was stuck and still not able to solve this issue.
 My focus changed to other project after that.
 
Wish you good luck on solving this problem, please let me know when you found a solution! Thanks.

**Comment by supercrazysam** [comment]
Hi Yu, since you found a working solution, if it is possible, would you mind sharing your working repo? Perhaps also the dataset as well?

**Comment by lovelyyunamars** [comment]
Hey, Sam,


Right now, I am trying 8 month data to see the solution is repeatable or not, if it works, I will share;)




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "CIA-Oceanix/GeoTrackNet"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年4月1日(星期四) 晚上6:10
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [CIA-Oceanix/GeoTrackNet] Using my own dataset. (#13)





 
Hi Yu, since you found a working solution, if it is possible, would you mind sharing your working repo? Perhaps also the dataset as well?

**Comment by lovelyyunamars** [comment]
Hi, Sam,

I tried my new datasets, it did not work well, similar bug happened about the D["seq"] can not be reshape to 4,  it is not repeatable, I am also stuck  there.

**Comment by lovelyyunamars** [comment]
> Hi Yu, since you found a working solution, if it is possible, would you mind sharing your working repo? Perhaps also the dataset as well?

Hi, Sam,
I am tried 8 month data to see the solution is repeatable or not, however it did not work well, similar bug happened about the D["seq"] can not be reshape to 4, it is not repeatable, I am also stuck there.
